vassal-app/src/main/java/VASSAL/script/ExpressionInterpreter.java

```diff
@@ -536,6 +579,190 @@ public Object countStack(String property, PropertySource ps) {
+   * @param property Property Name
+   * @param ps       GamePiece
+   * @return total
+   */
+  public Object sumAttachment(String attachment, String property, PropertySource ps) {
+    int result = 0;
+
+    ps = translatePiece(ps);
+
+    if (ps instanceof GamePiece) {
+      GamePiece p = Decorator.getOutermost((Decorator)ps);
+      while (p instanceof Decorator) {
+        if (p instanceof Attachment) {
+          final Attachment a = (Attachment)p;
+          if (a.getAttachName().equals(attachment)) {
+            for (final GamePiece target : a.getAttachList()) {
+              final Object prop = target.getProperty(property);
+              result += propValue(prop);
+            }
+          }
+        }
+        p = ((Decorator) p).getInner();
+      }
+    }
+    return result;
+  }
+
+
+  /**
+   * CountAttachment(property) function count the number of pieces
+   * attached to this piece via a named Attachment trait that
+   * *contain* the specified property
+   *
+   * @param attachment Attachment Name
+   * @param property Property Name
+   * @param ps       GamePiece
+   * @return total
+   */
+  public Object countAttachment(String attachment, String property, PropertySource ps) {
+    int result = 0;
+
+    ps = translatePiece(ps);
+
+    if (ps instanceof GamePiece) {
+      GamePiece p = Decorator.getOutermost((Decorator)ps);
+      while (p instanceof Decorator) {
+        if (p instanceof Attachment) {
+          final Attachment a = (Attachment)p;
+          if (a.getAttachName().equals(attachment)) {
+            for (final GamePiece target : a.getAttachList()) {
+              if ("".equals(property)) {
+                result++;
+              }
+              else {
+                final Object prop = target.getProperty(property);
+                result += propNonempty(prop);
+              }
+            }
+          }
+        }
+        p = ((Decorator) p).getInner();
+      }
+    }
+    return result;
+  }
+
+
+  /**
+   * CountAttachmentExpression(attachment, expression) function count the number of pieces
+   * attached to this piece via a named Attachment trait for which the expression is true
+   *
+   * @param attachment Attachment Name e.g. "MyTurrets"
+   * @param expression Expression e.g. " { Ammo > 2 } "
+   * @param ps       GamePiece
+   * @return total
+   */
+  public Object countAttachmentExpression(String attachment, String expression, PropertySource ps) {
+    int result = 0;
+
+    ps = translatePiece(ps);
+
+    if (ps instanceof GamePiece) {
+      GamePiece p = Decorator.getOutermost((Decorator)ps);
+      while (p instanceof Decorator) {
+        if (p instanceof Attachment) {
+          final Attachment a = (Attachment)p;
+          if (a.getAttachName().equals(attachment)) {
+            final String matchString = replaceDollarVariables(expression, ps);
+            final PieceFilter filter = matchString == null ? null : new PropertyExpression(unescape(matchString)).getFilter(ps);
```

Comment:
Should the above two lines be outside the loop?

Reply:
no -- p changes during the loop

Reply:
Oh I misunderstood which lines you meant -- I chose not to instantiate the beanshell until we actually find a matching attachment -- we should only ever find precisely 0 or 1

Reply:
In theory I guess I could put a "break' after finding a match, but that saves virtually nothing and conceivably some dodobrain could put 2 attachment traits with the same name

Reply:
I had not appreciated that detail about there never being more than one match. That might be worth a comment.

Reply:
added